Docs: Describe values under Extending Configuration Files (refs #6240)

pedrottimark opened the pull request:
For this docs counterpart to #6248 let’s aim for about 3 steps of critique:

1. Baseline with only **placeholder** content for `"eslint:all"`

    * Critique a new **structure** for Extending Configuration Files that is adapted from new structure under Options in rule docs: bulleted list of `extends` property values followed by level-3 headings with an example of configuration.

    * Critique reorganization and rewriting of existing **content**. Is it correct, clear, and complete in context? For example, this section is the target of the link about `"eslint:recommended"` on the rules index page. In particular, can whoever has expert knowledge ( @IanVS ? :) help me with the paragraph about “ESLint resolves a relative path…”

2. Baseline improvements plus **first draft** of new content about `"eslint:all"` including thoughts from #6240 and #6248 especially https://github.com/eslint/eslint/pull/6248#issuecomment-223628290

3. **Second draft** of new content about `"eslint:all"`

nzakas commented on docs/user-guide/configuring.md:
```diff
@@ -550,84 +550,125 @@ The complete configuration hierarchy, from highest precedence to lowest preceden
 ## Extending Configuration Files
 
-If you want to extend a specific configuration file, you can use the `extends` property and specify the path to the file. The path can be either relative or absolute.
+A configuration file can "extend" one or more "child" configurations to override a "base" set of rules.
```
Why the quotes around these terms? If the intent is to introduce new terms, then I'd suggest formally defining them in a terminology section.

nzakas commented on docs/user-guide/configuring.md:
```diff
@@ -550,84 +550,125 @@ The complete configuration hierarchy, from highest precedence to lowest preceden
+* all core rules
 
-The extended configuration provides base rules, which can be overridden by the configuration that references it. For example:
+If a "child" configuration has an `extends` property, ESLint extends the configuration recursively.
```
Definitely need to stop using quotes everywhere. :)

nzakas commented on docs/user-guide/configuring.md:
```diff
@@ -550,84 +550,125 @@ The complete configuration hierarchy, from highest precedence to lowest preceden
-1. JSON file
-1. JS file
-1. Shareable configuration package
+* the recommended rules
```
I think this list makes it more confusing rather than clarifying anything. The important thing to know about `extends` is that it can be a string or an array of strings, and that each string represents a shareable config to inherit from. What those strings represent doesn't make sense at this point in the conversation.

nzakas commented on docs/user-guide/configuring.md:
```diff
@@ -550,84 +550,125 @@ The complete configuration hierarchy, from highest precedence to lowest preceden
+* turn on additional rules
+
+
+### Extending the recommended rules
```
I think we need to work on the terminology for these headings. "Extending X" is a bit unclear to me. I'd rather saying "Using the recommended rules" and "Using a shareable configuration package"

pedrottimark commented:
Made changes from first critique, asks self “what was I thinking?”

Added example sentences to make config file format explicit and changed one to YAML.

Next up: first draft of content and example for `"eslint:all"`

pedrottimark commented:
Bring all feedback great and small to first draft of `"eslint:all"`

ilyavolodin commented:
Nice, LGTM. Should we put an even bigger disclaimer around `eslint:all` section. Using this will void your warranty and don't come crying to us?:-)

nzakas commented:
Yeah, I might just put this directly under the heading:

**Important:** This configuration changes with every minor and major release, and as such, is not recommended for production use. Use at your own risk.

pedrottimark commented:
Included **Important** paragraph after initial descriptive paragraph because it seemed to stand out a bit more there and because the following paragraph explains why.

Changed the first and last level-3 headings to give more descriptive anchors:
* Using `"eslint:recommended"`
* Using `"eslint:all"`

s/specify different options/override default options/

Changed from bulleted items to ordinary paragraphs. Rephrased to include original use case: especially if you rarely override options or disable rules.

nzakas commented:
LGTM

pedrottimark commented:
The first two paragraphs might work as a short summary for the version blog with link from `"eslint:all"` to the subheading in this doc.

The `extends` property value can be `"eslint:all"` to enable all core rules in the currently installed version of ESLint. The set of core rules can change at any minor or major version of ESLint.

**Important:** This configuration is **not recommended for production use** because it changes with every minor and major version of ESLint. Use at your own risk.

nzakas commented:
Yeah, I'll be sure to include that.
